sdk/lib/opentelemetry/sdk/trace/samplers/parent_consistent_probability_based.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Trace
+      module Samplers
+        # @api private
+        #
+        # The ParentConsistentProbabilityBased sampler is meant as an optional
+        # replacement for the ParentBased sampler. It is required to first validate
+        # the tracestate and then respect the sampled flag in the W3C traceparent.
+        #
+        # The ParentConsistentProbabilityBased Sampler constructor takes a single
+        # Sampler argument, which is the Sampler to use in case the
+        # ParentConsistentProbabilityBased Sampler is called for a root span.
+        class ParentConsistentProbabilityBased
+          include ConsistentProbabilityTraceState
+
+          def initialize(root)
+            @root = root
+          end
+
+          def ==(other)
+            @root == other.root
+          end
+
+          # @api private
+          #
+          # See {Samplers}.
+          def description
+            "ParentConsistentProbabilityBased{root=#{@root.description}}"
+          end
+
+          # @api private
+          #
+          # See {Samplers}.
+          def should_sample?(trace_id:, parent_context:, links:, name:, kind:, attributes:)
+            parent_span_context = OpenTelemetry::Trace.current_span(parent_context).context
+            if !parent_span_context.valid?
+              @root.should_sample?(trace_id: trace_id, parent_context: parent_context, links: links, name: name, kind: kind, attributes: attributes)
+            else
+              tracestate = validate_tracestate(parent_span_context)
```

Comment:
Interestingly, the [implementation for Go](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/efbcbfcb06541ef690398898f5d281068b92581b/samplers/probability/consistent/parent.go#L40-L67) always cleans up the `tracestate` even for root spans:
```go
	// Note: We do not check psc.IsValid(), i.e., we repair the tracestate
	// with or without a parent TraceId and SpanId.
	state := psc.TraceState()


	otts, err := parseOTelTraceState(state.Get(traceStateKey), psc.IsSampled())


	if err != nil {
		otel.Handle(err)
		value := otts.serialize()
		if len(value) > 0 {
			// Note: see the note in
			// "go.opentelemetry.io/otel/trace".TraceState.Insert(). The
			// error below is not a condition we're supposed to handle.
			state, _ = state.Insert(traceStateKey, value)
		} else {
			state = state.Delete(traceStateKey)
		}


		// Fix the broken tracestate before calling the delegate.
		params.ParentContext = trace.ContextWithSpanContext(params.ParentContext, psc.WithTraceState(state))
```

I couldn't find any requirement for this in the spec, which only states:

> It is required to first validate the `tracestate` and then respect the `sampled` flag in the W3C traceparent.